api/client/copy.go

```diff
@@ -0,0 +1,218 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/pkg/archive"
+)
+
+type copyDirection int
+
+const (
+	fromContainer copyDirection = (1 << iota)
+	toContainer
+	acrossContainers = fromContainer | toContainer
+)
+
+// CmdCp handles copying files or directories to/from/across containers.
+func (cli *DockerCli) CmdCp(args ...string) error {
+	cmd := cli.Subcmd("cp", "[SRC_CONTAINER:]SRC_PATH [DST_CONTAINER:]DST_PATH", "Copy files between containers or a local path", true)
+	if err := cmd.Parse(args); err != nil {
+		return nil
+	}
+
+	if cmd.NArg() != 2 {
+		cmd.Usage()
+		return nil
+	}
+
+	var (
+		srcPath      = cmd.Arg(0)
+		dstPath      = cmd.Arg(1)
+		srcContainer string
+		dstContainer string
+		direction    copyDirection
+	)
+
+	if strings.Contains(srcPath, ":") {
+		// Copy from a container.
+		direction |= fromContainer
+		srcInfo := strings.SplitN(srcPath, ":", 2)
+		srcContainer, srcPath = srcInfo[0], srcInfo[1]
+	}
+
+	if strings.Contains(dstPath, ":") {
+		// Copy to a container.
+		direction |= toContainer
+		dstInfo := strings.SplitN(dstPath, ":", 2)
+		dstContainer, dstPath = dstInfo[0], dstInfo[1]
+	}
+
+	switch direction {
+	case fromContainer:
+		// Copy from a container to a local path.
+		return cli.copyFromContainer(srcContainer, srcPath, dstPath)
+	case toContainer:
+		// Copy from a local path to a container.
+		return cli.copyToContainer(srcPath, dstContainer, dstPath)
+	case acrossContainers:
+		// Copy from one container to another.
+		return cli.copyAcrossContainers(srcContainer, srcPath, dstContainer, dstPath)
+	default:
+		// User didn't specify any container.
+		return fmt.Errorf("error: Wrong path format. Must specify a source and/or destination container")
+	}
+}
+
+type containerPathStat struct {
+	Name    string      `json:"name"`
+	AbsPath string      `json:"absPath"`
+	Size    int64       `json:"size"`
+	Mode    os.FileMode `json:"mode"`
+	Mtime   time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", path)
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
```

Comment:
> Is there any reason why we don't support something more natural, like:
>
>     GET /containers/1234234/files/myfile

The main reason is that the `engine` API that is used today makes it nearly impossible for the job to know that it's being used in the context of an HTTP request. It doesn't know anything about HTTP headers and is unable to set them. The job that I've written just writes an archive to its stdout.

What we *could* do though, is have this API endpoint handler run these two jobs one after the other, i.e, stat the path in the container, set the headers, then call out to the path archive job.

Does anyone know if we're able to handle HEAD/GET requests with the same handler function in  `api/server.go`?